### PR TITLE
adding llama code inference

### PR DESCRIPTION
**Pull request by HamidShojanazeri**
# What does this PR do?

This PR adds the code Llama examples for code completion and infilling.

- [ ] Test A
[Logs for infilling and completion ](https://gist.github.com/HamidShojanazeri/959e5b8f1ee96fe83870fdfb0b8ca23e)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: docs/inference.md [hunk 1]
```diff
@@ -41,6 +41,29 @@ model.resize_token_embeddings(model.config.vocab_size + 1)
...
 
+**Code Llama**
+
+Code llama was recently released with three flavors, base-model that support multiple programming languages, Python fine-tuned model and an instruction fine-tuned and aligned variation of Code Llama, please read more [here](https://ai.meta.com/blog/code-llama-large-language-model-coding/).
+
+Find the scripts to run Code Llama [here](../inference/code-llama/), where there are two examples of running code completion and infilling.
+
+**Note** Please find the right model on HF side [here](https://huggingface.co/codellama).
+
+To run the code completion example:
```
Comment (chauhang) [inline docs/inference.md:52]: Please add - currently requires installing transformers from source

File: inference/code-llama/code_completion_example.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    max_padding_length: int=None, # the max padding length to be used with tokenizer padding the prompts.
+    use_fast_kernels: bool = False, # Enable using SDPA from PyTroch Accelerated Transformers, make use Flash Attention and Xformer memory-efficient kernels
+    **kwargs
+):
+    if prompt_file is not None:
+        assert os.path.exists(
+            prompt_file
+        ), f"Provided Prompt file does not exist {prompt_file}"
+        with open(prompt_file, "r") as f:
+            user_prompt = "\n".join(f.readlines())
```
Comment (chauhang) [inline inference/code-llama/code_completion_example.py:45]: This is adding extra new lines in the prompts - which results in similar double new lines in the generated text, making it harder to read and use the generated code.

File: inference/code-llama/code_completion_example.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    top_p: float=1.0, # [optional] If set to float < 1, only the smallest set of most probable tokens with probabilities that add up to top_p or higher are kept for generation.
+    temperature: float=1.0, # [optional] The value used to modulate the next token probabilities.
+    top_k: int=50, # [optional] The number of highest probability vocabulary tokens to keep for top-k-filtering.
+    repetition_penalty: float=1.0, #The parameter for repetition penalty. 1.0 means no penalty.
+    length_penalty: int=1, #[optional] Exponential penalty to the length that is used with beam-based generation. 
+    enable_azure_content_safety: bool=False, # Enable safety check with Azure content safety api
+    enable_sensitive_topics: bool=False, # Enable check for sensitive topics using AuditNLG APIs
+    enable_salesforce_content_safety: bool=True, # Enable safety check with Salesforce safety flan t5
+    max_padding_length: int=None, # the max padding length to be used with tokenizer padding the prompts.
+    use_fast_kernels: bool = False, # Enable using SDPA from PyTroch Accelerated Transformers, make use Flash Attention and Xformer memory-efficient kernels
```
Comment (chauhang) [inline inference/code-llama/code_completion_example.py:37]: Please change default value to True

File: inference/code-llama/code_completion_example.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            print("Module 'optimum' not found. Please install 'optimum' it before proceeding.")
+
+    tokenizer = CodeLlamaTokenizer.from_pretrained(model_name)
+    tokenizer.add_special_tokens(
+        {
+         
+            "pad_token": "<PAD>",
+        }
+    )
+    model.resize_token_embeddings(model.config.vocab_size + 1)
```
Comment (chauhang) [inline inference/code-llama/code_completion_example.py:81]: Please verify this if needed in case of CodeLlama models. They already have extra "unused" tokens in the vocabulary/embeddings

File: inference/code-llama/code_infilling_example.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    **kwargs
+):
+    if prompt_file is not None:
+        assert os.path.exists(
+            prompt_file
+        ), f"Provided Prompt file does not exist {prompt_file}"
+        with open(prompt_file, "r") as f:
+            user_prompt = f.read()
+    elif not sys.stdin.isatty():
+        user_prompt = "\n".join(sys.stdin.readlines())
```
Comment (chauhang) [inline inference/code-llama/code_infilling_example.py:47]: This is adding extra new lines in the prompts - which results in similar double new lines in the generated text, making it harder to read and use the generated code.

File: docs/inference.md [hunk 6]
```diff
@@ -41,6 +41,29 @@ model.resize_token_embeddings(model.config.vocab_size + 1)
...
+
+```bash
+
+python code_completion_example.py --model_name MODEL_NAME  --prompt_file code_completion_prompt.txt --temperature 0.2 --top_p 0.9
+
+```
+
+To run the code infilling example:
```
Comment (chauhang) [inline docs/inference.md:60]: Add note to use the inflling supported models for this - Code Llama and Code Instruct models support infilling. Code Llama Python models do not support infilling

File: docs/inference.md [hunk 7]
```diff
@@ -41,6 +41,29 @@ model.resize_token_embeddings(model.config.vocab_size + 1)
 ```
 Padding would be required for batch inference. In this this [example](../inference/inference.py), batch size = 1 so essentially padding is not required. However,We added the code pointer as an example in case of batch inference.
 
+**Code Llama**
```
Comment (chauhang) [inline docs/inference.md:44]: Move this section after the Chat models section at the end

**Review by chauhang — CHANGES_REQUESTED** [review]
@HamidShojanazeri Thanks for getting this started. Please see the comments inline.

File: inference/code-llama/code_infilling_example.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            print("Module 'optimum' not found. Please install 'optimum' it before proceeding.")
+
+    tokenizer = CodeLlamaTokenizer.from_pretrained(model_name)
+    tokenizer.add_special_tokens(
+        {
+         
+            "pad_token": "<PAD>",
+        }
+    )
+    model.resize_token_embeddings(model.config.vocab_size + 1)
```
Comment (chauhang) [inline inference/code-llama/code_infilling_example.py:82]: Please confirm is the resize embeddings is needed for Code llama models as they have extra unused items in the vocab. At present generating a warning: 

>You are resizing the embedding layer without providing a `pad_to_multiple_of` parameter. This means that the new embedding dimension will be 32017. This might induce some performance reduction as *Tensor Cores* will not be available. For more details about this, or help on choosing the correct value for resizing, refer to this guide: https://docs.nvidia.com/deeplearning/performance/dl-performance-matrix-multiplication/index.html#requirements-tc

File: inference/code-llama/code_infilling_example.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                print(report)
+        print("Skipping the inference as the prompt is not safe.")
+        sys.exit(1)  # Exit the program with an error status
+        
+    batch = tokenizer(user_prompt, padding='max_length', truncation=True, max_length=max_padding_length, return_tensors="pt")
+    batch = {k: v.to("cuda") for k, v in batch.items()}
+    
+    start = time.perf_counter()
+    with torch.no_grad():
+        outputs = model.generate(
```
Comment (chauhang) [inline inference/code-llama/code_infilling_example.py:109]: For the Python model, generate is failing with error `Assertion srcIndex < srcSelectDimSize failed` in the Transfomers llama modelling code: https://gist.github.com/chauhang/833a9a624c99862997a5349f43881063

cc: @ArthurZucker, @faabian

File: inference/code-llama/code_completion_example.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    peft_model: str=None,
+    quantization: bool=False,
+    max_new_tokens =100, #The maximum numbers of tokens to generate
+    prompt_file: str=None,
+    seed: int=42, #seed value for reproducibility
+    do_sample: bool=True, #Whether or not to use sampling ; use greedy decoding otherwise.
+    min_length: int=None, #The minimum length of the sequence to be generated, input prompt + min_new_tokens
+    use_cache: bool=True,  #[optional] Whether or not the model should use the past last key/values attentions Whether or not the model should use the past last key/values attentions (if applicable to the model) to speed up decoding.
+    top_p: float=1.0, # [optional] If set to float < 1, only the smallest set of most probable tokens with probabilities that add up to top_p or higher are kept for generation.
+    temperature: float=1.0, # [optional] The value used to modulate the next token probabilities.
```
Comment (chauhang) [inline inference/code-llama/code_completion_example.py:29]: For the defaults use the same values as in the base codellama [generate](https://github.com/facebookresearch/codellama/blob/main/llama/generation.py#L116):

File: inference/code-llama/code_completion_example.py (hunk 3 above)
Comment (HamidShojanazeri) [reply inference/code-llama/code_completion_example.py]: done for both scripts

File: inference/code-llama/code_infilling_example.py (hunk 9 above)
Comment (HamidShojanazeri) [reply inference/code-llama/code_infilling_example.py:107]: This only happens for python model which is not supposed to be used for infilling added notes.

File: inference/code-llama/code_infilling_example.py (hunk 5 above)
Comment (HamidShojanazeri) [reply inference/code-llama/code_infilling_example.py]: removed this line `  user_prompt = "\n".join(sys.stdin.readlines())` as this is not feasible to format the code prompt from CLI.

File: inference/code-llama/code_completion_example.py (hunk 10 above)
Comment (HamidShojanazeri) [reply inference/code-llama/code_completion_example.py]: done

File: docs/inference.md (hunk 7 above)
Comment (chauhang) [reply docs/inference.md:51]: @HamidShojanazeri Will be better to move this Note right next to the Infilling example to remove confusion.
